node/src/crypto/hash.rs

```diff
@@ -172,6 +173,86 @@ impl From<Blake2bHash> for Digest {
     }
 }
 
+/// Hashes a pair of [Digest]s.
+pub fn hash_pair(hash1: &Digest, hash2: &Digest) -> Digest {
+    let mut to_hash = [0; Digest::LENGTH * 2];
+    to_hash[..Digest::LENGTH].copy_from_slice(&(hash1.to_array())[..]);
+    to_hash[Digest::LENGTH..].copy_from_slice(&(hash2.to_array())[..]);
+    hash(&to_hash)
+}
+
+/// Hash a [Vec<Digest>] into a single [Digest] by constructing a [Merkle tree][1].
+/// Reduces pairs of elements in the [Vec<Digest>] by repeatedly calling [hash_pair].
+/// This hash procedure is suited to hashing [BTree]s.
+///
+/// The pattern of hashing is as follows.  It is akin to [graph reduction][2]:
+///
+/// ```text
+/// a b c d e f
+/// |/  |/  |/
+/// g   h   i
+/// | /   /
+/// |/   /
+/// j   k
+/// | /
+/// |/
+/// l
+/// ```
+///
+/// Returns the empty [Digest] when the input is empty.
+///
+/// [1]: https://en.wikipedia.org/wiki/Merkle_tree
+/// [2]: https://en.wikipedia.org/wiki/Graph_reduction
+pub fn hash_vec_merkle_tree(vec: Vec<Digest>) -> Digest {
+    if vec.is_empty() {
+        return hash(&[]);
+    };
+    let mut vec = vec;
+    let mut k = vec.len();
+    while k != 1 {
+        let j = k / 2 + k % 2;
+        for i in 0..j {
+            if (2 * i + 1) > k {
+                vec[i] = vec[2 * i];
+            } else {
+                vec[i] = hash_pair(&vec[2 * i], &vec[2 * i + 1]);
+            }
+        }
+        k = j;
+    }
+    vec[0]
+}
```

Comment:
I just found today that there's a similar method in `itertools` lib for this: https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.tree_fold1

Reply:
Awesome, I didn't know that!
If we use that we should just have a unit test that makes sure the exact behavior of `tree_fold1` doesn't change, i.e. the shape of a tree with ≠ 2<sup>n</sup> leaves.

Reply:
I've made it use `itertools`, and there are tests checking hashing of vectors with odd and even number of elements, so if anything changes, they should start failing.